R/ranova.b.R

```diff
@@ -1,393 +1,1944 @@
+                            for (j in 1:nrow(mp)){
+                                A<-mp[j,]-center
+                                temp<-sum(A*B)*B/bot
+                                dis[j]<-sqrt(sum(temp^2))
+                            }
+
+                            dis.m=dis[1:nm]
+                            if(!MM){
+                                #temp<-idealf(dis)
+                                temp<-idealf(dis.m)
+                                dmat[,i]<-dis/(temp$qu-temp$ql)
+                            }
+
+                            if(MM)dmat[,i]<-dis/mad(dis.m)
+                        }
+                    }
+                    pdis<-apply(dmat,1,max,na.rm=na.rm)
+                    pdis=pdis[np1:nrow(mp)]
+                }
+
+                return(pdis)
+            }
+
+            idealf<-function(x,na.rm=FALSE){
+                # Compute the ideal fourths for data in x
+                if(na.rm)
+                    x<-x[!is.na(x)]
+
+                j<-floor(length(x)/4 + 5/12)
+                y<-sort(x)
+                g<-(length(x)/4)-j+(5/12)
+                ql<-(1-g)*y[j]+g*y[j+1]
+                k<-length(x)-j+1
+                qu<-(1-g)*y[k]+g*y[k-1]
+
+                rl<-list(ql=ql,qu=qu)
+
+                return(rl)
+            }
+
+            con<-as.matrix(con)
+            if(is.matrix(x) || is.data.frame(x))x=listm(x)
+            if(!is.list(x))stop("Data must be stored in list mode or in matrix mode.")
+
+            if(!is.na(grp)){  # Only analyze specified groups.
+                xx<-list()
+                for(i in 1:length(grp))
+                    xx[[i]]<-x[[grp[i]]]
+
+                x<-xx
+            }
+
+            J<-length(x)
+            mvec<-NA
+            nvec=NA
+
+            for(j in 1:J){
+                temp<-x[[j]]
+                if(na.rm)temp<-temp[!is.na(temp)] # Remove missing values.
+                x[[j]]<-temp
+                mvec[j]<-est(temp,...)
+                nvec[j]=length(temp)
+            }
+
+            Jm<-J-1
+            d<-ifelse(con==0,(J^2-J)/2,ncol(con))
+            if(sum(con^2)==0){
+                if(allp){
+                    con<-matrix(0,J,d)
+                    id<-0
+                    for (j in 1:Jm){
+                        jp<-j+1
+                        for (k in jp:J){
+                            id<-id+1
+                            con[j,id]<-1
+                            con[k,id]<-0-1
+                        }
+                    }
+                }
+
+                if(!allp){
+                    con<-matrix(0,J,Jm)
+                    for (j in 1:Jm){
+                        jp<-j+1
+                        con[j,j]<-1
+                        con[jp,j]<-0-1
+                    }
+                }
+            }
+
+            bvec<-matrix(NA,nrow=J,ncol=nboot)
+
+            # set seed of random number generator so that results can be duplicated.
+            if(SEED)set.seed(2)
```

Comment:
we shouldn't be setting the random seed ... if the user wants to set the seed, that's up to them, but we shouldn't be overriding the seed.

we run the risk that a random seed of 2 pushes the result into a really strange unlikely result, and that becomes our fault. where as if they get a crazy result from a random seed -- well that's just how these things work.

Reply:
What you say is important, but here we have two functions of R.R. Wilcox `medpb()` and `pbadepth()` (from the latest Rallfun-v40.txt).
They are fundamental to 3ways and have a **SEED argument** (=TRUE by default) which is left untouched and set to TRUE by default when called in` .mzMain3way()` and` .mzPH3way()`.
If I pass `SEED=FALSE`, `set.seed(2)` will not be applied, but the refresh recalculation will output different values and this is not what we are looking for.

I await (**when you have time**) your clarification to proceed, even with the rest.
TA

Reply:
> the refresh recalculation will output different values and this is not what we are looking for.

stochastic results changing with refresh is kinda what everyone does; i think it's what spss, jasp, jamovi all do.

i think it's the least bad option -- and it doesn't produce as much of a tech support burden (posts in the forum) as you might think.